contracts/serverless-v2/tree/EnvTreeUpgradeable.sol

```diff
@@ -0,0 +1,392 @@
+// SPDX-License-Identifier: MIT
+
+pragma solidity ^0.8.0;
+
+import "@openzeppelin/contracts-upgradeable/proxy/utils/Initializable.sol";
+
+/// @dev Contract implements an upgradeable version of tree which stores
+/// elements in order of insertion. When a element is added, it is added
+/// to the left most empty leaf in the tree. When an element is deleted,
+/// it is replaced with the element in the right most leaf in the tree.
+/// Each element in the tree stores the weight of all elements on the left
+/// and right side of the node.
+contract EnvTreeUpgradeable is Initializable {
+    /// @notice Struct that stores the value on the node and the sum of
+    /// weights on left and right side of the node.
+    /// @param value Value on the node
+    /// @param leftSum Sum of nodes on the left side of the current node
+    /// @param rightSum Sum of nodes on the right side of the current node
+    struct Node {
+        uint64 value;
+        uint64 leftSum;
+        uint64 rightSum;
+    }
+
+    struct Tree {
+        /// @notice Mapping of address of a node to it's index in nodes array
+        mapping(address => uint256) addressToIndexMap;
+        /// @notice Mapping of index in nodes array to address of the node
+        mapping(uint256 => address) indexToAddressMap;
+        /// @notice Array of nodes stored in the tree
+        Node[] nodes;
+    }
+
+    /// @custom:storage-location erc7201:marlin.oyster.storage.EnvTree
+    struct EnvTreeStorage {
+        /// @notice Tree data for each environment
+        mapping(uint8 => Tree) envTree;
+    }
+
+    // keccak256(abi.encode(uint256(keccak256("marlin.oyster.storage.EnvTree")) - 1)) & ~bytes32(uint256(0xff))
+    bytes32 private constant EnvTreeStorageLocation = 0x8731a315ce9ca0ea8593ae2136a4f652d3ece8e4802e703eae4270516b326400;
+
+    error TreeInvalidInitState();
+
+    function _getEnvTreeStorage() private pure returns (EnvTreeStorage storage $) {
+        assembly {
+            $.slot := EnvTreeStorageLocation
+        }
+    }
+
+    function __EnvTreeUpgradeable_init_unchained(uint8 _env) internal onlyInitializing {
```

Comment:
what is it supposed to do with _env? this init method is non-intuitive. Who would want to just init tree, give the option of multiple inits or just empty mapping of tree is fine, and new trees will be added later.

Reply:
updated to multiple init